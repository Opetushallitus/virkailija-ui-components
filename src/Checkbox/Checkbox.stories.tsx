import * as React from 'react';
import { storiesOf } from '@storybook/react';
import Checkbox from './index';
import { boolean } from '@storybook/addon-knobs';

const stories = storiesOf('Components/Checkbox', module);

stories.add('Default', () => {
  const disabled = boolean('Disabled', false) as boolean;
  const error = boolean('Error', false) as boolean;
  const checked = boolean('Checked', false) as boolean;
  const intederminate = boolean('Intederminate', false) as boolean;

  return (
    <Checkbox
      disabled={disabled}
      error={error}
      checked={checked}
      indeterminate={intederminate}
    >
      Checkbox
    </Checkbox>
  );
});
