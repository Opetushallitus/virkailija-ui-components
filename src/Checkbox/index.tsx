import * as React from 'react';
import styled, { css } from 'styled-components';
import { hideVisually } from 'polished';

import { disabledStyle } from '../system';

const CheckboxContainer = styled.div`
  position: relative;
  flex: 0;
  line-height: 0;
  top: 0.15em;
`;

const Icon = styled.svg`
  fill: none;
  stroke: white;
  stroke-width: 2px;
`;
const HiddenCheckbox = styled.input.attrs({ type: 'checkbox' })`
  ${hideVisually}
`;

const StyledCheckbox = styled.div<{ checked: boolean; error: boolean }>`
  display: inline-block;
  position: relative;
  width: 1em;
  height: 1em;
  background-color: white;
  border: 1px solid ${({ theme }) => theme.colors.inputBorder};
  box-shadow: 0 0 0 0 transparent;
  border-radius: 2px;
  transition: box-shadow 0.25s, border-color 0.25s, background-color 0.25s;

  &:hover, ${HiddenCheckbox}:focus + & {
    border-color: ${({ theme, error }) =>
      error ? theme.colors.danger.main : theme.colors.primary.main};
  }

  ${HiddenCheckbox}:focus + & {
    box-shadow: 0 0 0 3px ${({ theme }) => theme.colors.primary.focusOutline};
  }

  ${({ checked }) =>
    checked &&
    css`
      background-color: ${({ theme }) => theme.colors.primary.main};
      border-color: ${({ theme }) => theme.colors.primary.main};
    `}

  ${({ error }) =>
    error &&
    css`
      border-color: ${({ theme }) => theme.colors.danger.main};

      &:hover,
      ${HiddenCheckbox}:focus + & {
        border-color: ${({ theme }) => theme.colors.danger.main};
      }

      ${HiddenCheckbox}:focus + & {
        box-shadow: 0 0 0 3px ${({ theme }) => theme.colors.danger.focusOutline};
      }
    `}

  ${Icon} {
    visibility: ${props => (props.checked ? 'visible' : 'hidden')};
  }
`;

const Label = styled.label<{ disabled: boolean; error: boolean }>`
  cursor: pointer;
  font-family: ${({ theme }) => theme.fonts.main};
  font-size: 1rem;
  display: inline-flex;
  line-height: 1.5;
  color: ${({ theme }) => theme.colors.text.primary};
  align-items: flex-start;

  &:hover {
    & ${StyledCheckbox} {
      border-color: ${({ theme, error }) =>
        error ? theme.colors.danger.main : theme.colors.primary.main};
    }
  }

  ${disabledStyle}
`;

const LabelWrapper = styled.div`
  flex: 1;
  margin-left: 9px;
`;

const IndeterminateIndicator = styled.div`
  top: 50%;
  left: 50%;
  width: 0.5em;
  height: 0.5em;
  background-color: ${({ theme }) => theme.colors.primary.main};
  border-radius: 2px;
  position: absolute;
  transform: translate(-50%, -50%);
`;

type CheckboxBaseProps = {
  disabled?: boolean;
  checked?: boolean;
  children?: React.ReactNode;
  error?: boolean;
  className?: string;
  indeterminate?: boolean;
  fullWidth?: boolean;
};

export type CheckboxProps = CheckboxBaseProps &
  Omit<React.ComponentProps<typeof HiddenCheckbox>, keyof CheckboxBaseProps>;

const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
  (
    {
      className,
      checked = false,
      children,
      error = false,
      disabled = false,
      indeterminate = false,
      ...props
    },
    ref,
  ) => (
    <Label disabled={disabled} error={error} className={className}>
      <CheckboxContainer>
        <HiddenCheckbox
          checked={checked}
          disabled={disabled}
          ref={ref}
          {...props}
        />
        <StyledCheckbox checked={checked} error={error}>
          <Icon viewBox="0 0 24 24">
            <polyline points="20 6 9 17 4 12" />
          </Icon>
          {indeterminate && !checked ? <IndeterminateIndicator /> : null}
        </StyledCheckbox>
      </CheckboxContainer>
      {children && <LabelWrapper>{children}</LabelWrapper>}
    </Label>
  ),
);

export default Checkbox;
