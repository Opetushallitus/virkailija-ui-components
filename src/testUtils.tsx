import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { createTheme } from './createTheme';

const theme = createTheme();

const Providers = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const dummyDiv = document.createElement('div');
dummyDiv.setAttribute('data-nothing', 'true');

// Return rootElement from render() for easier snapshot testing
const customRender = (ui: React.ReactElement, options: any = {}) => {
  const result = render(ui, { wrapper: Providers, ...options });
  return {
    ...result,
    rootElement: result.container.firstChild || dummyDiv,
  };
};

export * from '@testing-library/react';

export { customRender as render };
