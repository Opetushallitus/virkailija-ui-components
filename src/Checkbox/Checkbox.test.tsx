import React from 'react';
import { screen } from '@testing-library/react';
import Checkbox from './index';
import { render, fireEvent } from '../testUtils';

test('Renders correctly with defaults', () => {
  const { rootElement } = render(
    <Checkbox onChange={() => {}}>Checkbox</Checkbox>,
  );
  expect(rootElement).toMatchSnapshot();
});

test('Clicking calls onChange-callback', () => {
  const changeHandler = jest.fn();
  const { rootElement } = render(
    <Checkbox onChange={changeHandler}>Checkbox</Checkbox>,
  );
  fireEvent.click(rootElement);
  expect(changeHandler).toBeCalledTimes(1);
  expect(rootElement).toMatchSnapshot();
});

test('Checked-attribute should be set when checked', () => {
  const changeHandler = jest.fn();
  render(
    <Checkbox onChange={changeHandler} checked={true}>
      Checkbox
    </Checkbox>,
  );
  const checkbox = screen.getByLabelText('Checkbox');

  expect(checkbox).toHaveAttribute('checked');
});
