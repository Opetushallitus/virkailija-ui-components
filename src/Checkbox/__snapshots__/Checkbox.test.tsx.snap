// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clicking calls onChange-callback 1`] = `
.c1 {
  position: relative;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  line-height: 0;
  top: 0.15em;
}

.c7 {
  fill: none;
  stroke: white;
  stroke-width: 2px;
}

.c3 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c5 {
  display: inline-block;
  position: relative;
  width: 1em;
  height: 1em;
  background-color: white;
  border: 1px solid #ced4da;
  box-shadow: 0 0 0 0 transparent;
  border-radius: 4px;
  -webkit-transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
  transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
}

.c5:hover,
.c2:focus + .c5 {
  border-color: #0a789c;
}

.c2:focus + .c4 {
  box-shadow: 0 0 0 3px rgba(10,120,156,0.2);
}

.c5 .c6 {
  visibility: hidden;
}

.c0 {
  cursor: pointer;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1.5;
  color: #666;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0:hover .c4 {
  border-color: #0a789c;
}

.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-left: 9px;
}

<label
  class="c0"
>
  <div
    class="c1"
  >
    <input
      class="c2 c3"
      type="checkbox"
    />
    <div
      class="c4 c5"
    >
      <svg
        class="c6 c7"
        viewBox="0 0 24 24"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
  </div>
  <div
    class="c8"
  >
    Checkbox
  </div>
</label>
`;

exports[`Renders correctly with defaults 1`] = `
.c1 {
  position: relative;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  line-height: 0;
  top: 0.15em;
}

.c7 {
  fill: none;
  stroke: white;
  stroke-width: 2px;
}

.c3 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c5 {
  display: inline-block;
  position: relative;
  width: 1em;
  height: 1em;
  background-color: white;
  border: 1px solid #ced4da;
  box-shadow: 0 0 0 0 transparent;
  border-radius: 4px;
  -webkit-transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
  transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
}

.c5:hover,
.c2:focus + .c5 {
  border-color: #0a789c;
}

.c2:focus + .c4 {
  box-shadow: 0 0 0 3px rgba(10,120,156,0.2);
}

.c5 .c6 {
  visibility: hidden;
}

.c0 {
  cursor: pointer;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1.5;
  color: #666;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0:hover .c4 {
  border-color: #0a789c;
}

.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-left: 9px;
}

<label
  class="c0"
>
  <div
    class="c1"
  >
    <input
      class="c2 c3"
      type="checkbox"
    />
    <div
      class="c4 c5"
    >
      <svg
        class="c6 c7"
        viewBox="0 0 24 24"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
  </div>
  <div
    class="c8"
  >
    Checkbox
  </div>
</label>
`;
